tcmur_cmd_handler.c

```diff
@@ -628,26 +584,58 @@ static int handle_passthrough(struct tcmu_device *dev,
 	return ret;
 }
 
-/*
- * try to passthrough the command if handler supports command passthrough.
- * note that TCMU_NOT_HANDLED is returned when a tcmur handler does not
- * handle a passthrough command, but since we call ->handle_cmd via
- * async_call_command(), ->handle_cmd can finish in the callers context
- * (asynchronous handler) or work queue context (synchronous handlers),
- * thus we'd need to check if ->handle_cmd handled the passthough command
- * here as well as in handle_passthrough_cbk().
- */
+bool tcmur_handler_is_passthrough_only(struct tcmur_handler *rhandler)
+{
+	if (rhandler->write || rhandler->read || rhandler->flush)
+		return false;
+
+	return true;
+}
+
 int tcmur_cmd_handler(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
 {
-	int ret;
+	int ret = TCMU_NOT_HANDLED;
 	struct tcmulib_handler *handler = tcmu_get_dev_handler(dev);
 	struct tcmur_handler *rhandler = handler->hm_private;
+	uint8_t *cdb = cmd->cdb;
+	struct iovec *iovec = cmd->iovec;
+	size_t iov_cnt = cmd->iov_cnt;
+	uint32_t block_size = tcmu_get_dev_block_size(dev);
+	off_t offset = block_size * tcmu_get_lba(cdb);
 
-	if (rhandler->handle_cmd) {
-		ret = handle_passthrough(dev, rhandler, cmd);
-		if ((ret == TCMU_ASYNC_HANDLED) || (ret != TCMU_NOT_HANDLED))
-			return ret;
```

Comment:
wondering why you dropped 'command passthough' mechanism out. If only ceph can offload command and others can't, we should only passthrough the command to rbd driver and allow others enjoy generic handling. Besides, you are making tcmu-runner more and more ceph centeric, losing modularity. for e.g., transition_state and report_state should be in the upper geneic layer.  Store drier should only concentrate on the store details, not on the iscsi protocol details.